### PR TITLE
Add Rusanov boundary correction for the relativistic Euler system

### PR DESCRIPTION
**Pull request by fmahebert**
## Proposed changes

Depends on PRS #2967 and #2969 

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

File: tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/BoundaryCorrections/Test_Rusanov.cpp [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+template <size_t Dim, typename EosType>
+void test(const size_t num_pts, const EosType& equation_of_state) {
+  tuples::TaggedTuple<hydro::Tags::EquationOfState<EosType>> volume_data{
+      equation_of_state};
+  tuples::TaggedTuple<
+      helpers::Tags::Range<hydro::Tags::RestMassDensity<DataVector>>,
+      helpers::Tags::Range<hydro::Tags::SpecificInternalEnergy<DataVector>>,
+      helpers::Tags::Range<hydro::Tags::SpecificEnthalpy<DataVector>>,
+      helpers::Tags::Range<hydro::Tags::SpatialVelocity<DataVector, Dim>>>
+      ranges{std::array<double, 2>{{1e-6, 1.0}},
```
Comment (nilsdeppe) [inline tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/BoundaryCorrections/Test_Rusanov.cpp]: I'm not sure what the lower range should be. `1.0e-30` still seems reasonable to test.

Also `1e-6` -> `1.0e-6`

**Review by nilsdeppe — COMMENTED** [review]
Overall looks good, one small nitpick. Feel free to squash immediately

**Comment by fmahebert** [comment]
Rebased on develop now that no longer dependent.

I've also squashed in a few small changes:
- changed bounds/notation as per Nils's review
- cleaned up computation of max char speeds in the python comparison implementation

**Review by nilsdeppe — DISMISSED** [review]

File: tests/Unit/Helpers/Evolution/DiscontinuousGalerkin/BoundaryCorrections.hpp [hunk 2]
```diff
@@ -290,6 +291,27 @@ void test_boundary_correction_conservation_impl(
         make_not_null(&exterior_unit_normal_covector),
         make_not_null(&exterior_unit_normal_vector),
         get<inv_spatial_metric>(exterior_fields_on_face));
+
+    // if dg_package_data depends on the (not inverted) spatial metric, we
+    // also adjust this. Note we that for testing purposes, the metric and
```
Comment (nilsvu) [inline tests/Unit/Helpers/Evolution/DiscontinuousGalerkin/BoundaryCorrections.hpp]: <s>we</s>

File: src/Evolution/Systems/RelativisticEuler/Valencia/BoundaryCorrections/Rusanov.cpp [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& normal_covector,
+    const tnsr::I<DataVector, Dim, Frame::Inertial>& /*normal_vector*/,
+    const std::optional<tnsr::I<DataVector, Dim, Frame::Inertial>>&
+    /*mesh_velocity*/,
+    const std::optional<Scalar<DataVector>>& normal_dot_mesh_velocity,
+    const EquationsOfState::EquationOfState<true, ThermodynamicDim>&
+        equation_of_state) const noexcept {
+  {
+    // First compute v^2 and cs^2
+    Scalar<DataVector>& spatial_velocity_squared =
```
Comment (nilsvu) [inline src/Evolution/Systems/RelativisticEuler/Valencia/BoundaryCorrections/Rusanov.cpp:75]: (optional) `auto&`, and would be helpful to add a comment that the two return variables are used as memory buffer

File: tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/BoundaryCorrections/Rusanov.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+import numpy as np
+
+import Evolution.Systems.RelativisticEuler.Valencia.TestFunctions as valencia
+
+
+def dg_package_data_tilde_d(
+    tilde_d, tilde_tau, tilde_s, flux_tilde_d, flux_tilde_tau, flux_tilde_s,
+    lapse, shift, spatial_metric, rest_mass_density, specific_internal_energy,
+    specific_enthalpy, spatial_velocity, normal_covector, normal_vector,
+    mesh_velocity, normal_dot_mesh_velocity, use_polytropic_fluid):
```
Comment (nilsvu) [inline tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/BoundaryCorrections/Rusanov.py:13]: (optional) you can abbreviate all of these unused args with `dg_package_data_tilde_d(tilde_d, *args, **kwargs):`

File: tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/BoundaryCorrections/Rusanov.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    interior_normal_dot_flux_tilde_s, interior_abs_char_speed,
+    exterior_tilde_d, exterior_tilde_tau, exterior_tilde_s,
+    exterior_normal_dot_flux_tilde_d, exterior_normal_dot_flux_tilde_tau,
+    exterior_normal_dot_flux_tilde_s, exterior_abs_char_speed,
+    use_strong_form):
+    if use_strong_form:
+        return -0.5 * (interior_normal_dot_flux_tilde_d +
+                       exterior_normal_dot_flux_tilde_d) - 0.5 * np.maximum(
+                           interior_abs_char_speed, exterior_abs_char_speed
+                       ) * (exterior_tilde_d - interior_tilde_d)
```
Comment (nilsvu) [inline tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/BoundaryCorrections/Rusanov.py]: (optional) math often formats nicer in Python when you wrap it in parens: `return (-0.5 * ...)`. Could help with a bunch of expressions in this file.

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/BoundaryCorrections/Rusanov.py (hunk 4 above)
Comment (fmahebert) [reply tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/BoundaryCorrections/Rusanov.py:13]: This is new info to me, so thanks for the pointer! I think I will leave as-is for consistency with elsewhere, though?

File: tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/BoundaryCorrections/Rusanov.py (hunk 5 above)
Comment (fmahebert) [reply tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/BoundaryCorrections/Rusanov.py]: Doing this here gave a small but appreciable formatting improvement, so I went with it.

**Comment by fmahebert** [comment]
Fixups posted

File: tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/BoundaryCorrections/Rusanov.py (hunk 4 above)
Comment (nilsvu) [reply tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/BoundaryCorrections/Rusanov.py:13]: Up to you, but I don't think we have to prioritize consistency here. Also this is already used in `FishboneMoncriefDisk.py`, `KomissarovShock.py` and `FirstOrderInternalPenalty.py` (disclaimer: I wrote 2/3 of those).

**Review by nilsvu — COMMENTED** [review]
I'm happy with squashing the fixups. Also feel free to squash if you decide to go with the `*args` syntax.

**Comment by nilsdeppe** [comment]
fixups look fine. I personally have started avoiding `auto` more because having the type can help with readability and also helps with auto completion in code editors, but I'm fine either way here.

**Comment by fmahebert** [comment]
Okay, since @nilsdeppe agrees with me on preferring explicit types, and @nilsleiffischer's request was optional, I reverted back to writing `Scalar<DataVector>` on the two temporaries.

Squashed the fixups and rebased on develop.

**Review by nilsdeppe — APPROVED** [review]

**Review by nilsvu — APPROVED** [review]
Re `auto`: We should probably follow https://google.github.io/styleguide/cppguide.html#Type_deduction more closely :) (use `auto` only where it helps readability)

**Comment by nilsvu** [comment]
(Note that our current guidelines contradict the Google style guide, so we may want to change them https://spectre-code.org/code_review_guide.html)

**Pull request merged**